### ISSUE TITLE
Azure Storage - Got invalid blocklist when flushing user

### ISSUE THREAD

**Issue by andersosthus** [grafana/loki#5235]
**Describe the bug**
I restarted one of my ingesters, and during startup I saw the following in the logs (see below).
I have not studied the code so I don't know if there is any parallelism here, but one reason for this could be if the blocklist call is done before the blobs are finished uploading.

**Expected behavior**

**Environment:**
 - Infrastructure: Kubernetes
 - Deployment tool: Kustomize
 - Loki version: 2.4.1

**Screenshots, Promtail config, or terminal output**
Ingester log:
```
level=error ts=2022-01-25T21:46:04.146509336Z caller=flush.go:221 org_id=ORGID msg="failed to flush user" err="
-> github.com/Azure/azure-storage-blob-go/azblob.newStorageError, /src/loki/vendor/github.com/Azure/azure-storage-blob-go/azblob/zc_storage_error.go:42
===== RESPONSE ERROR (ServiceCode=InvalidBlockList) =====
Description=The specified block list is invalid.
RequestId:b73c827d-201e-0056-1234-128ecb000000
Time:2022-01-25T21:46:04.1461630Z, Details:
Code: InvalidBlockList
PUT https://MYACCOUNT.blob.core.windows.net/lokichunks/ORGID/ec8b5c649be1c7a1-17e930cf0dd-17e93181f1b-a8b90bbc?comp=blocklist&timeout=10
Authorization: REDACTED
Content-Length: [128]
Content-Type: [application/xml]
User-Agent: [Azure-Storage/0.13 (go1.17.2; linux)]
X-Ms-Blob-Cache-Control: []
X-Ms-Blob-Content-Disposition: []
X-Ms-Blob-Content-Encoding: []
X-Ms-Blob-Content-Language: []
X-Ms-Blob-Content-Type: []
X-Ms-Client-Request-Id: [d76a549d-8e5c-4508-797f-930151255a4b]
X-Ms-Date: [Tue, 25Jan 2022 21:46:04 GMT]
X-Ms-Version: [2019-12-12]
--------------------------------------------------------------------------------
RESPONSE Status: 400 The specified block list is invalid.
Content-Length: [221]
Content-Type: [application/xml]
Date: [Tue, 25 Jan 2022 21:46:03 GMT]
Server: [Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0]
X-Ms-Client-Request-Id: [d76a549d-8e5c-4508-797f-930151255a4b]
X-Ms-Error-Code: [InvalidBlockList]
X-Ms-Request-Id: [b73c827d-201e-0056-1234-128ecb000000]
X-Ms-Version: [2019-12-12]"

level=error ts=2022-01-25T21:46:04.147182246Z caller=flush.go:221 org_id=ORGID msg="failed to flush user" err="
-> github.com/Azure/azure-storage-blob-go/azblob.newStorageError, /src/loki/vendor/github.com/Azure/azure-storage-blob-go/azblob/zc_storage_error.go:42
===== RESPONSE ERROR (ServiceCode=InvalidBlockList) =====
Description=The specified block list is invalid.
RequestId:25f5d9d9-e01e-0076-4934-12f56c000000
Time:2022-01-25T21:46:04.1471551Z, Details:
Code: InvalidBlockList
PUT https://MYACCOUNT.blob.core.windows.net/lokichunks/ORGID/c2a7d8c98583f4fd-17e93186689-17e93186689-9d79a7e6?comp=blocklist&timeout=10
Authorization: REDACTED
Content-Length: [128]
Content-Type: [application/xml]
User-Agent: [Azure-Storage/0.13 (go1.17.2; linux)]
X-Ms-Blob-Cache-Control: []
X-Ms-Blob-Content-Disposition: []
X-Ms-Blob-Content-Encoding: []
X-Ms-Blob-Content-Language: []
X-Ms-Blob-Content-Type: []
X-Ms-Client-Request-Id: [31e68ff9-7b82-435a-600f-c3b5b934b0cd]
X-Ms-Date: [Tue, 25Jan 2022 21:46:04 GMT]
X-Ms-Version: [2019-12-12]
--------------------------------------------------------------------------------
RESPONSE Status: 400 The specified block list is invalid.
Content-Length: [221]
Content-Type: [application/xml]
Date: [Tue, 25 Jan 2022 21:46:03 GMT]
Server: [Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0]
X-Ms-Client-Request-Id: [31e68ff9-7b82-435a-600f-c3b5b934b0cd]
X-Ms-Error-Code: [InvalidBlockList]
X-Ms-Request-Id: [25f5d9d9-e01e-0076-4934-12f56c000000]
X-Ms-Version: [2019-12-12]"
```

**Comment by cyriltovena** [comment]
Could be an azure client quirk. The flush will be retried so it should be fine, let us know if this is happening too often.

**Comment by andersosthus** [comment]
Not sure what counts as "too often", but looking back at the logs, it seems to have started yesterday evening when I restarted the Ingesters, and since then there are 13 entries in the log with this issue. But before I restarted the Ingesters yesterday, I don't see any at all.

I'll keep an eye on it in the coming days and see if it continues or not.

**Comment by ovidiubuligan** [comment]
@andersosthus Have you figgured out the source of those InvalidBlockList logs  ?  I am seeing similar problem with grafana tempo index file .

**Comment by andersosthus** [comment]
@ovidiubuligan No, I didn't draw any conclusions on this, and I haven't really watched for those issues after. But I can scan my logs and see, both for Loki and Tempo, if I still get them.

**Comment by aslafy-z** [comment]
Un-stale please

**Comment by aslafy-z** [comment]
Not stale please

**Comment by slim-bean** [comment]
The root cause of this issue appears to be explained here:

https://github.com/Azure/azure-sdk-for-go/issues/18490#issuecomment-1170806877

And sure seems a lot like "wontfix" from the Azure sdk as they point the issue out to a decision made by a vendored google library they use to write to their own storage 🙃 

I don't believe this error is overly impactful, at least in our environments we don't see it that often and the failed chunk push requests will be retried so there should be no data loss or impact, just some annoying logs and increased retries.

I guess to make this go away we'd have to change our implementation some how, it's not clear to me at a cursory glance what exactly we would need to do...

**Comment by slim-bean** [comment]
I did take a poke at our implementation, and the situation is a little different than I just described

https://github.com/grafana/loki/blob/7cc6aa402166c64e2b6592cb82a6e605743f08a9/pkg/storage/chunk/client/azure/blob_storage_client.go#L233-L247

It looks like we directly call the `UploadStreamToBlockBlob` method.

I'm not sure I see how this results in the issue as described or really what a better approach would be here, anyone with more experience with Azure that can offer suggestions here would be greatly appreciated.

**Comment by jbizcloud** [comment]
All, any news on this topic? It is impacting my client so badly.

**Comment by sivesh1989** [comment]
Hi,

          We also observe same error while running in loki 2.8 with bolt DB shipper as well as TSDB and also observe drop in lines/sec processed. Also tried with azure premium storage account and same error was observed.

Load: 25k lines/sec

 msg="failed to flush" err="failed to flush chunks: store put chunk: -> github.com/Azure/azure-storage-blob-go/azblob.newStorageError, /drone/src/vendor/github.com/Azure/azure-storage-blob-go/azblob/zc_storage_error.go:42\n===== RESPONSE ERROR (ServiceCode=InvalidBlockList) =====\nDescription=The specified block list is invalid.\nRequestId:d8f1bfed-b01e-0088-0f4b-b9245b000000\nTime:2023-07-18T07:41:17.4840736Z, Details: \n Code: InvalidBlockList\n PUT [.blob.core.windows.net/loki?comp=blocklist&timeout=31\n] Authorization: REDACTED\n Content-Length: [233]\n Content-Type: [application/xml]\n User-Agent: [Azure-Storage/0.14 (go1.20.4 linux)]\n X-Ms-Blob-Cache-Control: []\n X-Ms-Blob-Content-Disposition: []\n X-Ms-Blob-Content-Encoding: []\n X-Ms-Blob-Content-Language: []\n X-Ms-Blob-Content-Type: []\n X-Ms-Client-Request-Id: [7e8e7591-228d-4fce-7562-2b9b5cdee1bc]\n X-Ms-Date: [Tue, 18 Jul 2023 07:41:17 GMT]\n X-Ms-Version: [2020-04-08]\n --------------------------------------------------------------------------------\n RESPONSE Status: 400 The specified block list is invalid.\n Content-Length: [221]\n Content-Type: [application/xml]\n Date: [Tue, 18 Jul 2023 07:41:17 GMT]\n Server: [Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0]\n X-Ms-Client-Request-Id: [7e8e7591-228d-4fce-7562-2b9b5cdee1bc]\n X-Ms-Error-Code: [InvalidBlockList]\n X-Ms-Request-Id: [d8f1bfed-b01e-0088-0f4b-b9245b000000]\n X-Ms-Version: [2020-04-08]\n\n\n,

Screenshot:
![image](https://github.com/grafana/loki/assets/47570718/94ca3158-dd0a-4c2e-b458-d20c71a2a225)

Regards,
Sivesh Kumar R

**Comment by kunalmehta-eve** [comment]
Hey @slim-bean ,  were we able to fix this by any chance ?

**Comment by JStickler** [comment]
Closing as stale. This issue was opened against a version of Loki that is no longer supported. If this issues still exists in more recent versions of Loki, please leave a comment and we can reopen the issue.
